app/queues/dns/dns-worker.server.ts

```diff
@@ -0,0 +1,142 @@
+  'dns-queue',
+  async (job) => {
+    logger.debug('worker started');
+    let { step = Step.Route53Change } = job.data;
+
+    while (step !== Step.Finished) {
+      switch (step) {
+        case Step.Route53Change: {
+          const changeId = await executeChange({
+            ...job.data,
+          });
+          await job.update({ ...job.data, changeId, step: Step.WaitOnRoute53Change });
+        }
+
+        case Step.WaitOnRoute53Change: {
+          const dnsStatus = await waitOnChange(job.data.changeId);
+          await job.update({ ...job.data, dnsStatus, step: Step.SyncChangeToDatabase });
+        }
+
+        case Step.SyncChangeToDatabase: {
+          const { dnsRecordId, dnsOperation, dnsStatus } = job.data;
+          await syncChange(dnsRecordId, dnsOperation, dnsStatus);
+          step = Step.Finished;
+          return step;
+        }
+
+        default: {
+          throw new UnrecoverableError(`invalid step: ${step}`);
+        }
+      }
+    }
+  },
+  {
+    connection: redis,
+    limiter: {
+      max: 5,
+      duration: 1_000,
+    },
+  }
+);
+
+async function executeChange({ dnsOperation, username, type, fqdn, value }: DnsChangeData) {
+  logger.debug(
+    `Route53 DNS change: ${dnsOperation} ${fqdn}=${value} with type=${type} for user=${username}`
+  );
+
+  try {
+    switch (dnsOperation) {
+      case 'create':
+        return createDnsRecord(username, type, fqdn, value);
+      case 'update':
+        return upsertDnsRecord(username, type, fqdn, value);
+      case 'delete':
+        return deleteDnsRecord(username, type, fqdn, value);
+      default:
+        throw new UnrecoverableError(`Invalid DNS Operation: ${dnsOperation}`);
+    }
+  } catch (error) {
+    logger.warn(
+      `Route53 DNS change error: unable to ${dnsOperation} ${fqdn} (value=${value}) for user=${username}`,
+      error
+    );
+    throw error;
+  }
+}
+
+async function waitOnChange(changeId?: string) {
+  logger.debug(`Checking status of Route53 DNS Change ${changeId}`);
+
+  // If we don't get back a Change ID, it means there's nothing to
+  // wait on and everything is already in sync.
+  if (!changeId) {
+    return 'INSYNC';
+  }
+
+  // Otherwise, see if Route53 has finished syncing this change
+  const status = await getChangeStatus(changeId);
+
+  if (status === 'PENDING') {
+    throw new Error(`Route53 DNS Change ${changeId} still pending`);
+  }
+
+  return status;
+}
+
+async function syncChange(id: DnsRecord['id'], dnsOperation: DnsOperation, dnsStatus?: string) {
+  logger.debug(`Syncing DNS record id=${id} with DNS status=${dnsStatus} to database`);
+
+  try {
+    switch (dnsOperation) {
+      case 'delete':
+        if (dnsStatus === 'INSYNC') {
+          return deleteDnsRecordById(id);
+        }
+        break;
+      default:
+        return updateDnsRecordById(id, {
+          status: dnsStatus === 'INSYNC' ? DnsRecordStatus.active : DnsRecordStatus.error,
```

Comment:
I'm not sure if this is expected. I returned `undefined` when creating a record, it updates the DB as `active`. I know we are going to have reconciler, but the user will see my record is available right away. Probably we should add `switch` by job type in `waitOnChange` to handle `create` differently?

Reply:
This is exactly what we want.  When we `create` the DB record will be available right away, but `pending`.  When it finishes, we update to be either `active` or `error` and the UI changes to match.